src/goto-analyzer/static_verifier.cpp

```diff
@@ -300,32 +325,127 @@ bool static_verifier(
+          ++pass;
+          break;
+        case static_verifier_resultt::TRUE:
+          ++pass;
+          break;
+        case static_verifier_resultt::FALSE:
+          ++fail;
+          break;
+        case static_verifier_resultt::UNKNOWN:
+          ++unknown;
+          break;
+        default:
+          UNREACHABLE;
+        }
       }
       else
       {
-        result.status = static_verifier_resultt::UNKNOWN;
-        ++unknown;
+        // Multiple traces, verify against each one
+        std::size_t unreachable_traces = 0;
+        std::size_t true_traces = 0;
+        std::size_t false_traces = 0;
+        std::size_t unknown_traces = 0;
+
+        for(const auto &trace_ptr : trace_set)
+        {
+          auto dp = ai.abstract_state_before(trace_ptr);
+
+          result.status = check_assertion(*dp, e, ns);
+          switch(result.status)
+          {
+          case static_verifier_resultt::BOTTOM:
+            ++unreachable_traces;
+            break;
+          case static_verifier_resultt::TRUE:
+            ++true_traces;
+            break;
+          case static_verifier_resultt::FALSE:
+            ++false_traces;
+            break;
+          case static_verifier_resultt::UNKNOWN:
+            ++unknown_traces;
+            break;
+          default:
+            UNREACHABLE;
+          }
+        }
+
+        // Join the results
+        if(unknown_traces != 0)
+        {
+          // If any trace is unknown, the final result must be unknown
+          result.status = static_verifier_resultt::UNKNOWN;
+          ++unknown;
+        }
+        else
+        {
+          if(false_traces == 0)
+          {
+            // Definitely true; the only question is how
+            ++pass;
+            if(true_traces == 0)
+            {
+              // Definitely not reachable
+              INVARIANT(
+                unreachable_traces == trace_set.size(),
+                "All traces must not reach the assertion");
+              result.status = static_verifier_resultt::BOTTOM;
+            }
+            else
+            {
+              // At least one trace (may) reach it.
+              // All traces that reach it are safe.
+              result.status = static_verifier_resultt::TRUE;
+            }
+          }
+          else
+          {
+            // At lease one trace (may) reach it and it is false on that trace
+            if(true_traces == 0)
+            {
+              // All traces that (may) reach it are false
+              ++fail;
+              result.status = static_verifier_resultt::FALSE;
+            }
+            else
+            {
+              // The awkward case, there are traces that (may) reach it and
+              // some are true, some are false.  It is not entirely fair to say
+              // "FAILURE (if reachable)" because it's a bit more complex than
+              // that, "FAILURE (if reachable via a particular trace)" would be
```

Comment:
So normally `FALSE` just meant "definitely fails and is definitely reachable" IIRC? Isn't this breaking convention?

Reply:
It is not really possible or meaningful to follow BMC style conventions in an abstract interpreter.  In BMC we are always dealing with the possibility that there are executions that we haven't explored, so "TRUE / SAFE" comes with an implicit "(in all the traces we have explored)" but on the plus side "FALSE / UNSAFE" is a definite thing.

In the AI case, "TRUE / SAFE" really is in *all* traces but "FALSE / UNSAFE (if reachable)" is the best we can manage.

I'm not sure I see an easy way of resolving this; it is certainly more fundamental than this PR.